### PR TITLE
XD-2942: Add ftp source.

### PR DESCRIPTION
**Pull request by ericbottard**
Based off work submitted by fmarchand (https://github.com/spring-projects/spring-xd/pull/1554)
Added e2e test, refactored module options some more

### CODE REVIEW

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSourceOptionsMetadata.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    public void setTmpFileSuffix(String tmpFileSuffix) {
+        this.tmpFileSuffix = tmpFileSuffix;
+    }
+
+    @Min(0)
+    public int getFixedRate() {
+        return fixedRate;
+    }
+
+    @ModuleOption("fixed delay in SECONDS to poll the remote directory")
```
Comment (fmarchand) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSourceOptionsMetadata.java]: I think fixed-rate for polling consumer is in milliseconds by default. isn't it ?

File: modules/source/ftp/config/ftp.xml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        	<beans:property name="password" value="${password:#{null}}"/>
+        	<beans:property name="clientMode" value="${clientMode}"/>
+	</beans:bean>
+
+	<int-ftp:inbound-channel-adapter auto-startup="false" channel="files" session-factory="ftpSessionFactory"
+		filename-pattern="${filenamePattern}" remote-directory="${remoteDir}" local-directory="${localDir}"
+		remote-file-separator="${remoteFileSeparator}" preserve-timestamp="${preserveTimestamp}"
+		auto-create-local-directory="${autoCreateLocalDir}"
+		temporary-file-suffix="${tmpFileSuffix}" delete-remote-files="${deleteRemoteFiles}">
+		<poller fixed-rate="${fixedRate}" />
```
Comment (garyrussell) [inline modules/source/ftp/config/ftp.xml]: New pollers should use `PeriodicTriggerMixin` - see https://jira.spring.io/browse/XD-2846

File: spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSourceOptionsMetadata.java (hunk 1 above)
Comment (garyrussell) [inline spring-xd-dirt/src/main/java/org/springframework/xd/dirt/modules/metadata/FtpSourceOptionsMetadata.java]: True, but see my other comment about `PeriodicTriggerMixin`.

**Comment by ericbottard** [comment]
Rebased, added PeriodicTriggerMixin

File: modules/source/ftp/config/ftp.xml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	<channel id="output" />
+
+	<channel id="files" />
+
+	<beans:beans profile="use-ref">
+		<bridge input-channel="files" output-channel="output"/>
+	</beans:beans>
+
+	<beans:beans profile="use-contents">
```
Comment (garyrussell) [inline modules/source/ftp/config/ftp.xml]: Hmmm - why did we add these profiles? 

I am not saying they are bad but it seems like gold plating and we should really have the `sftp` and `ftp` sources similarly configured - I think that means we should add these options to SFTP (or at least open a story for that).

File: spring-xd-shell/src/test/java/org/springframework/xd/shell/command/AbstractStreamIntegrationTest.java [hunk 4]
```diff
@@ -28,17 +28,7 @@
...
-import org.springframework.xd.test.fixtures.FileSource;
-import org.springframework.xd.test.fixtures.JdbcSink;
-import org.springframework.xd.test.fixtures.JdbcSource;
-import org.springframework.xd.test.fixtures.MailSink;
-import org.springframework.xd.test.fixtures.NonPollingImapSource;
-import org.springframework.xd.test.fixtures.PollingMailSource;
-import org.springframework.xd.test.fixtures.TailSource;
-import org.springframework.xd.test.fixtures.TcpSink;
-import org.springframework.xd.test.fixtures.TcpSource;
+import org.springframework.xd.test.fixtures.*;
```
Comment (garyrussell) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/AbstractStreamIntegrationTest.java]: Have we switched back to wildcard imports?

Comment (ericbottard) [inline spring-xd-shell/src/test/java/org/springframework/xd/shell/command/AbstractStreamIntegrationTest.java]: Nope, but I switched laptops :) Will fix

Warning, one of your favorite discussions ahead: Why aren't the formatting options into VCS again?

File: modules/source/ftp/config/ftp.xml (hunk 3 above)
Comment (ericbottard) [inline modules/source/ftp/config/ftp.xml]: Are your arguing about the use of profiles, or the ref/contents feature itself? While testing this, it quickly became obvious to me that having just the ref is hard to deal with, so I added this (which is on par with the `file` source).
So if you're discussing the latter, then yes I agree every source should behave the same

Comment (garyrussell) [inline modules/source/ftp/config/ftp.xml]: Yes, it's the latter - I realize it came from the `file` source, but it's not consistent with the implementation there (although I prefer this over the router there - and it's __completely missing__ from sftp :smile: )

**Comment by ericbottard** [comment]
Created https://jira.spring.io/browse/XD-3013 to capture the `--ref=true|false` option for stfp

**Comment by ericbottard** [comment]
Rebased, fixed imports.

**Comment by garyrussell** [comment]
Merged as ed27e32103205c149b420b593ad06309cfd41071 with minor polish spaces->tabs etc.

**Pull request closed without merging**